File: examples/excel-data-validations.ts
/**
 * 例子（Epic-58）：演示 Excel 数据验证（下拉列表 + 数值范围）。
 *
 * 跑法：
 *   bun run examples/excel-data-validations.ts <output.xlsx>
 */

import { StringValue, UInt32Value } from "../src/element/index.js";
import {
  Cell,
  CellValue,
  Row,
  SheetData,
  SpreadsheetDocument,
  addCellListValidation,
  addCellRangeValidation,
  getCellValidations,
} from "../src/excel/index.js";

async function main(): Promise<void> {
  const [outputPath] = process.argv.slice(2);
  if (outputPath === undefined) {
    process.stderr.write("usage: excel-data-validations <output.xlsx>\n");
    process.exit(2);
  }

  const doc = SpreadsheetDocument.create();
  const ws = doc.workbookPart?.worksheetParts[0]?.worksheet;
  if (ws === undefined) throw new Error("expected default worksheet");
  const sheetData = ws.firstChild(SheetData);
  if (sheetData === undefined) throw new Error("expected default SheetData");

  // 添加标题行
  const headerRow = new Row();
  headerRow.rowIndex = UInt32Value.parse("1");
  for (const [ref, text] of [
    ["A1", "水果选择（A 列）"],
    ["B1", "数量（B 列）"],
    ["C1", "价格（C 列，小数）"],
  ] as const) {
    const cell = new Cell();
    cell.cellReference = StringValue.parse(ref);
    cell.dataType = StringValue.parse("str");
    const v = new CellValue();
    v.text = text;
    cell.appendChild(v);
    headerRow.appendChild(cell);
  }
  sheetData.appendChild(headerRow);

  // 添加数据行（留空让用户填）
  for (let i = 2; i <= 11; i += 1) {
    const row = new Row();
    row.rowIndex = UInt32Value.parse(String(i));
    sheetData.appendChild(row);
  }

  // A 列：下拉列表验证（水果选择）
  addCellListValidation(ws, "A2:A11", {
    values: ["苹果", "香蕉", "樱桃", "橙子", "西瓜"],
    promptTitle: "请选择水果",
    prompt: "从下拉列表中选择一种水果",
    errorTitle: "无效输入",
    error: "请从列表中选择水果",
  });

  // B 列：整数范围验证（数量 1~1000）
  addCellRangeValidation(ws, "B2:B11", {
    type: "whole",
    min: 1,
    max: 1000,
    promptTitle: "请输入数量",
    prompt: "请输入 1 到 1000 之间的整数",
    errorTitle: "无效数量",
    error: "数量必须是 1 到 1000 之间的整数",
  });

  // C 列：小数范围验证（价格 0.01~9999.99）
  addCellRangeValidation(ws, "C2:C11", {
    type: "decimal",
    min: 0.01,
    max: 9999.99,
    promptTitle: "请输入价格",
    prompt: "请输入 0.01 到 9999.99 之间的价格",
  });

  const validations = getCellValidations(ws);
  process.stdout.write(`数据验证规则数量: ${validations.length}\n`);
  for (const v of validations) {
    process.stdout.write(
      `  ${v.sqref}: type=${v.type}, formula1=${v.formula1}, formula2=${v.formula2}\n`,
    );
  }

  await doc.saveAsAsync(outputPath);
  process.stdout.write(`已写入 ${outputPath}\n`);
}

main().catch((err) => {
  process.stderr.write(`failed: ${(err as Error).stack ?? err}\n`);
  process.exit(1);
});

File: tools/cross-sdk-verify/replicas/ExcelDataValidations.cs
// Replica of examples/excel-data-validations.ts
// Creates an xlsx with header row + 10 data rows, 3 data validations:
//   A2:A11 - list (fruits)
//   B2:B11 - whole number range 1-1000
//   C2:C11 - decimal range 0.01-9999.99

using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Spreadsheet;

namespace CrossSdkVerify.Replicas;

public static class ExcelDataValidations
{
    public static void Run(string outputPath)
    {
        using var doc = SpreadsheetDocument.Create(outputPath,
            DocumentFormat.OpenXml.SpreadsheetDocumentType.Workbook);

        var wbPart = doc.AddWorkbookPart();
        wbPart.Workbook = new Workbook();

        var wsPart = wbPart.AddNewPart<WorksheetPart>();
        var sheetData = new SheetData();
        var ws = new Worksheet(sheetData);

        var sheets = wbPart.Workbook.AppendChild(new Sheets());
        sheets.Append(new Sheet
        {
            Id = wbPart.GetIdOfPart(wsPart),
            SheetId = 1,
            Name = "Sheet1"
        });

        // Header row
        var headerRow = new Row { RowIndex = 1 };
        headerRow.Append(MakeStrCell("A1", "水果选择（A 列）"));
        headerRow.Append(MakeStrCell("B1", "数量（B 列）"));
        headerRow.Append(MakeStrCell("C1", "价格（C 列，小数）"));
        sheetData.Append(headerRow);

        // Empty data rows 2-11
        for (uint i = 2; i <= 11; i++)
            sheetData.Append(new Row { RowIndex = i });

        // Data validations
        var dvList = new DataValidations();

        // A2:A11 - list validation
        dvList.Append(new DataValidation
        {
            Type = DataValidationValues.List,
            SequenceOfReferences = new ListValue<StringValue> { InnerText = "A2:A11" },
            Formula1 = new Formula1("\"苹果,香蕉,樱桃,橙子,西瓜\""),
            ShowInputMessage = true,
            ShowErrorMessage = true,
            PromptTitle = new StringValue("请选择水果"),
            Prompt = new StringValue("从下拉列表中选择一种水果"),
            ErrorTitle = new StringValue("无效输入"),
            Error = new StringValue("请从列表中选择水果"),
        });

        // B2:B11 - whole number 1-1000
        dvList.Append(new DataValidation
        {
            Type = DataValidationValues.Whole,
            Operator = DataValidationOperatorValues.Between,
            SequenceOfReferences = new ListValue<StringValue> { InnerText = "B2:B11" },
            Formula1 = new Formula1("1"),
            Formula2 = new Formula2("1000"),
            ShowInputMessage = true,
            ShowErrorMessage = true,
            PromptTitle = new StringValue("请输入数量"),
            Prompt = new StringValue("请输入 1 到 1000 之间的整数"),
            ErrorTitle = new StringValue("无效数量"),
            Error = new StringValue("数量必须是 1 到 1000 之间的整数"),
        });

        // C2:C11 - decimal 0.01-9999.99
        dvList.Append(new DataValidation
        {
            Type = DataValidationValues.Decimal,
            Operator = DataValidationOperatorValues.Between,
            SequenceOfReferences = new ListValue<StringValue> { InnerText = "C2:C11" },
            Formula1 = new Formula1("0.01"),
            Formula2 = new Formula2("9999.99"),
            ShowInputMessage = true,
            ShowErrorMessage = true,
            PromptTitle = new StringValue("请输入价格"),
            Prompt = new StringValue("请输入 0.01 到 9999.99 之间的价格"),
        });

        ws.Append(dvList);
        wsPart.Worksheet = ws;
        wbPart.Workbook.Save();
    }

    private static Cell MakeStrCell(string cellRef, string value) =>
        new Cell
        {
            CellReference = cellRef,
            DataType = CellValues.String,
            CellValue = new CellValue(value)
        };
}
